src/phys/dca_step/cluster_solver/shared_tools/accumulation/tp/tp_accumulator_kernels.cu

```diff
@@ -0,0 +1,434 @@
+  // id_j is a linearized index of b3, b4, k2, k_ex.
+  const int id_j = blockIdx.y * blockDim.y + threadIdx.y;
+  // id_z is a linearized index of k_ex, w_ex.
+  const int id_z = blockIdx.z * blockDim.z + threadIdx.z;
+  if (id_i >= size || id_j >= size || id_z >= nw_exchange * nk_exchange)
+    return;
+
+  // Unroll id_i and id_j.
+  const int step2 = nb * nb;
+  const int step1 = step2 * nk;
+  auto get_indices = [=](int id, int& b1, int& b2, int& k, int& w) {
+    w = id / step1;
+    id -= w * step1;
+    k = id / step2;
+    id -= k * step2;
+    b2 = id / nb;
+    b1 = id - nb * b2;
+  };
+  int w1, w2, k1, k2, b1, b2, b3, b4;
+  get_indices(id_i, b1, b2, k1, w1);
+  get_indices(id_j, b3, b4, k2, w2);
+
+  // Unroll the exchange index id_z = k_ex + nk_exchange * w_ex.
+  const int w_ex = id_z / nk_exchange;
+  const int k_ex = id_z - w_ex * nk_exchange;
+
+  CudaComplex<Real> contribution;
+  const Real factor = 0.5 * sign;
+  const int no = nk * nb;
+  auto cond_conj = [](const CudaComplex<Real> a, const bool cond) { return cond ? conj(a) : a; };
+
+  // Compute the contribution.
+  switch (type) {
+    case PARTICLE_HOLE_TRANSVERSE: {
+      int w1_a(w1), w2_a(w2);
+      const bool conj_a = helper.extendWIndices(w1_a, w2_a);
+      const int i_a = b1 + nb * k1 + no * w1_a;
+      const int j_a = b4 + nb * k2 + no * w2_a;
+
+      int w1_b(helper.addWex(w2, w_ex));
+      int w2_b(helper.addWex(w1, w_ex));
+      const bool conj_b = helper.extendWIndices(w1_b, w2_b);
+      const int i_b = b2 + nb * helper.addKex(k2, k_ex) + no * w1_b;
+      const int j_b = b3 + nb * helper.addKex(k1, k_ex) + no * w2_b;
+
+      const CudaComplex<Real> Ga_1 = cond_conj(G_up[i_a + ldgu * j_a], conj_a);
+      const CudaComplex<Real> Gb_1 = cond_conj(G_down[i_b + ldgd * j_b], conj_b);
+
+      const CudaComplex<Real> Ga_2 = cond_conj(G_down[i_a + ldgd * j_a], conj_a);
+      const CudaComplex<Real> Gb_2 = cond_conj(G_up[i_b + ldgu * j_b], conj_b);
+
+      contribution = (Ga_1 * Gb_1 + Ga_2 * Gb_2) * (-factor);
+    } break;
+
+    case PARTICLE_HOLE_MAGNETIC: {
+      int w1_a(w1), w2_a(w2);
+      const bool conj_a = helper.extendWIndices(w1_a, w2_a);
+      const int i_a = b1 + nb * k1 + no * w1_a;
+      const int j_a = b4 + nb * k2 + no * w2_a;
+
+      int w1_b(helper.addWex(w2, w_ex));
+      int w2_b(helper.addWex(w1, w_ex));
+      const bool conj_b = helper.extendWIndices(w1_b, w2_b);
+      const int i_b = b2 + nb * helper.addKex(k2, k_ex) + no * w1_b;
+      const int j_b = b3 + nb * helper.addKex(k1, k_ex) + no * w2_b;
+
+      const CudaComplex<Real> Ga_1 = cond_conj(G_up[i_a + ldgu * j_a], conj_a);
+      const CudaComplex<Real> Gb_1 = cond_conj(G_up[i_b + ldgu * j_b], conj_b);
+
+      const CudaComplex<Real> Ga_2 = cond_conj(G_down[i_a + ldgd * j_a], conj_a);
+      const CudaComplex<Real> Gb_2 = cond_conj(G_down[i_b + ldgd * j_b], conj_b);
+
+      contribution = (Ga_1 * Gb_1 + Ga_2 * Gb_2) * (-factor);
+    }
+      {
```

Comment:
Here and again at line 299 - there is a chunk of (pasted) duplicate code. Can you confirm that this is intentional and not a mistaken cut'n'paste error.